Cics PUT CONTAINER improvements

mm-broadcom opened the pull request:
## How Has This Been Tested?

Created both valid and invalid instances and added tests verifying as such.

## Checklist:
- [x] Each of my commits contains one meaningful change
- [x] I have performed rebase of my branch on top of the development
- [x] I have performed a self-review of my own code
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] I have checked my code and corrected any misspellings

mm-broadcom committed:
Initial grammar

mm-broadcom committed:
Add checks for grammar

mm-broadcom committed:
Add tests for grammar

mm-broadcom committed:
Fixed checkstyle error

mm-broadcom committed:
Add PREPEND to lexer, keywords and the regex

mm-broadcom committed:
Added note about PREPEND's availability.

Added a note informing the user that PREPEND is only available in CICS 6.2 or later.

mm-broadcom committed:
Add PUT64 keyword support

Add support for PUT64 which has the same exact grammar as PUT CONTAINER (Channel).

mm-broadcom commented:
@KutluOzel-b Could you review this?

KutluOzel-b commented on server/engine/src/main/java/org/eclipse/lsp/cobol/implicitDialects/cics/utility/CICSPutContainerOptionsCheckUtility.java:
```diff
@@ -0,0 +1,81 @@
+        checkDuplicates(ctx);
+    }
+
+    private void checkBTS(CICSParser.Cics_put_container_btsContext ctx) {
```
Add FROM as mandatory

mm-broadcom replied:
Resolved in latest commit.

KutluOzel-b commented on server/engine/src/main/java/org/eclipse/lsp/cobol/implicitDialects/cics/utility/CICSPutContainerOptionsCheckUtility.java:
```diff
@@ -0,0 +1,81 @@
+        checkMutuallyExclusiveOptions("ACTIVITY, ACQACTIVITY, PROCESS or ACQPROCESS", ctx.ACTIVITY(), ctx.ACQACTIVITY(), ctx.PROCESS(), ctx.ACQPROCESS());
+    }
+
+    private void checkChannel(CICSParser.Cics_put_container_channelContext ctx) {
```
Add FROM as mandatory

mm-broadcom replied:
Resolved in latest commit.

KutluOzel-b commented on server/engine/src/test/java/org/eclipse/lsp/cobol/usecases/TestCicsPutContainerStatement.java:
```diff
@@ -0,0 +1,98 @@
+ * <p>This class tests all variations of the CONVERSE command found in the link above.
+ */
+public class TestCicsPutContainerStatement {
+    // Main Building Blocks
+    private static final String BASE_TEXT =
+            "       IDENTIFICATION DIVISION.\n"
+                    + "       PROGRAM-ID. ABCDEF.\n"
+                    + "       DATA DIVISION.\n"
+                    + "       WORKING-STORAGE SECTION.\n"
+                    + "       PROCEDURE DIVISION.\n"
+                    + "            EXEC CICS \n"
+                    + "            END-EXEC.";
```
```suggestion

```

KutluOzel-b commented on server/engine/src/test/java/org/eclipse/lsp/cobol/usecases/TestCicsPutContainerStatement.java:
```diff
@@ -0,0 +1,98 @@
+    private static final String PUT_BTS_INVALID = "PUT CONTAINER(123) ACTIVITY(123) {ACQACTIVITY|errorOne} FROM(123) FLENGTH(123)";
+    private static final String PUT_CHANNEL_INVALID = "PUT CONTAINER(123) CHANNEL(123) FROM(123) FLENGTH(123) BIT {DATATYPE|errorOne}(123) APPEND";
+
+    // Utility Functions
+    private static void noErrorTest(String newCommand) {
+        UseCaseEngine.runTest(getTestString(newCommand), ImmutableList.of(), ImmutableMap.of());
+    }
+
+    private static String getTestString(String newCommand) {
+        List<String> instances = Arrays.asList(newCommand.split("\\s"));
+        instances.replaceAll(String.join("", Collections.nCopies(12, " "))::concat);
+        ArrayList<String> base = new ArrayList<String>(Arrays.asList(BASE_TEXT.split("\n")));
+        base.addAll(base.size() - 1, instances);
+        return String.join("\n", base);
+    }
```
```suggestion

```

KutluOzel-b commented on server/engine/src/test/java/org/eclipse/lsp/cobol/usecases/TestCicsPutContainerStatement.java:
```diff
@@ -0,0 +1,98 @@
+    // Test Functions
+    @Test
+    void testBTSValid() {
+        noErrorTest(PUT_BTS_VALID_1);
+        noErrorTest(PUT_BTS_VALID_2);
```
```suggestion
        CICSTestUtils.noErrorTest(PUT_BTS_VALID_1);
        CICSTestUtils.noErrorTest(PUT_BTS_VALID_2);
```

KutluOzel-b commented on server/engine/src/test/java/org/eclipse/lsp/cobol/usecases/TestCicsPutContainerStatement.java:
```diff
@@ -0,0 +1,98 @@
+
+    @Test
+    void testChannelValid() {
+        noErrorTest(PUT_CHANNEL_VALID_1);
+        noErrorTest(PUT_CHANNEL_VALID_2);
```
```suggestion
        CICSTestUtils.noErrorTest(PUT_CHANNEL_VALID_1);
        CICSTestUtils.noErrorTest(PUT_CHANNEL_VALID_2);
```

KutluOzel-b commented on server/engine/src/test/java/org/eclipse/lsp/cobol/usecases/TestCicsPutContainerStatement.java:
```diff
@@ -0,0 +1,98 @@
+        noErrorTest(PUT_CHANNEL_VALID_2);
+    }
+
+    // Invalid Tests
```
Add tests case for FROM is mandatory

mm-broadcom replied:
Resolved in latest commit.

mm-broadcom committed:
Update server/engine/src/test/java/org/eclipse/lsp/cobol/usecases/TestCicsPutContainerStatement.java

Co-authored-by: Kutlu <104970275+KutluOzel-b@users.noreply.github.com>

mm-broadcom committed:
Update server/engine/src/test/java/org/eclipse/lsp/cobol/usecases/TestCicsPutContainerStatement.java

Co-authored-by: Kutlu <104970275+KutluOzel-b@users.noreply.github.com>

mm-broadcom committed:
Update server/engine/src/test/java/org/eclipse/lsp/cobol/usecases/TestCicsPutContainerStatement.java

Co-authored-by: Kutlu <104970275+KutluOzel-b@users.noreply.github.com>

mm-broadcom committed:
Update server/engine/src/test/java/org/eclipse/lsp/cobol/usecases/TestCicsPutContainerStatement.java

Co-authored-by: Kutlu <104970275+KutluOzel-b@users.noreply.github.com>

mm-broadcom committed:
Add FROM check, test, fix imports

KutluOzel-b reviewed (approved):
LGTM 👍

amanPras commented on server/engine/src/main/java/org/eclipse/lsp/cobol/implicitDialects/cics/utility/CICSOptionsCheckUtility.java:
```diff
@@ -147,6 +150,9 @@ public CICSOptionsCheckUtility(DialectProcessingContext context, List<SyntaxErro
     optionsMap.put(
         CICSWriteOptionsCheckUtility.RULE_INDEX,
         new CICSWriteOptionsCheckUtility(context, errors));
+    optionsMap.put(
```
looks unrelated.

slavek-kucera replied:
Duplicate.

amanPras commented:
LGTM! waiting for the merge conflicts to be resolved

KutluOzel-b commented on server/engine/src/test/java/org/eclipse/lsp/cobol/usecases/TestCicsPutContainerStatement.java:
```diff
@@ -0,0 +1,81 @@
+public class TestCicsPutContainerStatement {
+
+    // Test Strings
+    private static final String PUT_BTS_VALID_1 = "PUT CONTAINER(123) ACTIVITY(123) FROM(123) FLENGTH(123)";
+    private static final String PUT_BTS_VALID_2 = "PUT CONTAINER(123) ACQACTIVITY FROM(123)";
+
+    private static final String PUT_CHANNEL_VALID_1 = "PUT CONTAINER(123) CHANNEL(123) FROM(123) FLENGTH(123) BIT FROMCCSID(123) APPEND";
+    private static final String PUT_CHANNEL_VALID_2 = "PUT CONTAINER(123) FROM(123)";
+
+    private static final String PUT_BTS_INVALID = "PUT CONTAINER(123) ACTIVITY(123) {ACQACTIVITY|errorOne} FROM(123) FLENGTH(123)";
+    private static final String PUT_CHANNEL_INVALID = "PUT CONTAINER(123) CHANNEL(123) FROM(123) FLENGTH(123) BIT {DATATYPE|errorOne}(123) APPEND";
+    private static final String PUT_BTS_INVALID_2 = "PUT CONTAINER(123) {ACQACTIVITY|errorOne}";
```
After latest decision, please use variables as suboperand

slavek-kucera commented on server/engine/src/main/antlr4/org/eclipse/lsp/cobol/implicitDialects/cics/CICSLexer.g4:
```diff
@@ -660,6 +661,7 @@ PURGEABILITY : P U R G E A B I L I T Y;
 PURGEABLE : P U R G E A B L E;
 PUSH : P U S H;
 PUT : P U T;
+PUT64: P U T '6' '4';
```
Should these new tokens be added to the `cicsLexerDefinedVariableUsageTokens` rule?

mm-broadcom replied:
Done.

mm-broadcom committed:
Add terms to cicsLexerDefinedVariableUsage rule

mm-broadcom commented on server/engine/src/main/java/org/eclipse/lsp/cobol/implicitDialects/cics/utility/CICSPutContainerOptionsCheckUtility.java:
```diff
@@ -0,0 +1,84 @@
+        checkHasMandatoryOptions(ctx.FROM(), ctx, "FROM");
+    }
+
+    private void checkChannel(CICSParser.Cics_put_container_channelContext ctx) {
```
That was added to the documentation _**after**_ this PR was originally made. I can go back and add support for it.

mm-broadcom replied:
Added in most recent commit.

mm-broadcom committed:
Added check for APPEND or PREPEND

slavek-kucera commented on server/engine/src/main/antlr4/org/eclipse/lsp/cobol/implicitDialects/cics/CICSParser.g4:
```diff
@@ -2120,6 +2121,7 @@ ABCODE
  | PREDICATE
  | PREFIX
  | PREPARE
+ | PREPEND
```
I think PUT64 should be in the list as well.

mm-broadcom replied:
Fixed in latest commit.

mm-broadcom committed:
Add PUT64, remove duplicate optionsMap entry

slavek-kucera commented on server/engine/src/main/java/org/eclipse/lsp/cobol/implicitDialects/cics/utility/CICSOptionsCheckUtility.java:
```diff
@@ -157,17 +160,14 @@ public CICSOptionsCheckUtility(DialectProcessingContext context, List<SyntaxErro
     optionsMap.put(
-        CICSWriteOptionsCheckUtility.RULE_INDEX,
-        new CICSWriteOptionsCheckUtility(context, errors));
+        CICSWSAEPRUtility.RULE_INDEX,
```
`CICSWriteOptionsCheckUtility` was removed from the list in the last commit.

mm-broadcom replied:
Resolved in latest commit.

slavek-kucera commented on server/engine/src/main/java/org/eclipse/lsp/cobol/implicitDialects/cics/utility/CICSPutContainerOptionsCheckUtility.java:
```diff
@@ -0,0 +1,85 @@
+    private static final Map<Integer, ErrorSeverity> DUPLICATE_CHECK_OPTIONS =
+        new HashMap<Integer, ErrorSeverity>() {
+            {
+                put(CICSLexer.PUT, ErrorSeverity.ERROR);
```
`PUT` is not an operand.

mm-broadcom replied:
Fixed in latest commit.

mm-broadcom committed:
Re-add Write to OptionsMap

Re-add Write to the OptionsMap. Was mistakenly removed during fixing conflicts while doing the rebase.

mm-broadcom committed:
Add to Duplicate Check Options

mm-broadcom committed:
Add missing semi-colon

Add missing semi-colon. Forgot that VS Code's "Sort Lines Ascending" has the tendency to omit the last character in the resultant selection.

mm-broadcom committed:
Change tests to use variables

slavek-kucera commented on server/engine/src/main/antlr4/org/eclipse/lsp/cobol/implicitDialects/cics/CICSParser.g4:
```diff
@@ -679,10 +679,9 @@ cics_purge: PURGE cics_handle_response? MESSAGE cics_handle_response?;
-cics_put_bts: (ACTIVITY cics_data_value | ACQACTIVITY | PROCESS | ACQPROCESS | FROM cics_data_area | FLENGTH cics_data_value | cics_handle_response)+;
-cics_put_channel: (CHANNEL cics_data_value | FROM cics_data_area | FLENGTH cics_data_value | BIT | DATATYPE cics_cvda |
-                  CHAR | FROMCCSID cics_data_value | FROMCODEPAGE cics_data_value | cics_handle_response)+;
+cics_put_container: ((PUT CONTAINER cics_data_value cics_put_container_bts) | (PUT|PUT64) CONTAINER cics_data_value cics_put_container_channel);
+cics_put_container_bts: ((ACQACTIVITY | PROCESS | ACQPROCESS) | (ACTIVITY | FLENGTH) cics_data_value | (FROM) cics_data_area | cics_handle_response)+;
+cics_put_container_channel: ((BIT | CHAR | APPEND | PREPEND) | (CHANNEL | FLENGTH | FROMCCSID | FROMCODEPAGE) cics_data_value | FROM cics_data_area | DATATYPE cics_cvda | cics_handle_response)+;
```
The CONTAINER does not have to be mentioned first.

slavek-kucera commented on server/engine/src/main/java/org/eclipse/lsp/cobol/implicitDialects/cics/utility/CICSPutContainerOptionsCheckUtility.java:
```diff
@@ -0,0 +1,93 @@
+    private static final Map<Integer, ErrorSeverity> DUPLICATE_CHECK_OPTIONS =
+        new HashMap<Integer, ErrorSeverity>() {
+            {
+                put(CICSLexer.ACQACTIVITY, ErrorSeverity.ERROR);
```
ACQACTIVITY duplicates produce warnings only.

slavek-kucera commented on server/engine/src/main/java/org/eclipse/lsp/cobol/implicitDialects/cics/utility/CICSPutContainerOptionsCheckUtility.java:
```diff
@@ -0,0 +1,93 @@
+        checkHasMandatoryOptions(ctx.FROM(), ctx, "FROM");
+    }
+
+    private void checkChannel(CICSParser.Cics_put_container_channelContext ctx) {
```
We should also issue the error about PUT64 not being available in COBOL.

mm-broadcom replied:
Added in latest commit.

mm-broadcom committed:
Make grammar flexible, add checks, fix test

mm-broadcom committed:
Add warning for PUT64 usage in COBOL

Added a warning/error when PUT64 is used, inform user that it is for Assembly only, add a test for it.

slavek-kucera committed:
undo merge artifacts
